examples/inference/distributed/florence2.py

```diff
@@ -0,0 +1,203 @@
+            collation_fn=partial(collate_fn, processor=processor),
+        )
+    )
+    dataloader = wds.WebLoader(
+        dataset,
+        batch_size=None,
+        num_workers=num_workers,
+        pin_memory=True,
+        persistent_workers=True,
+    )
+
+    def save_results(output_queue: queue.Queue, output_dir: pathlib.Path, processor):
+        while True:
+            try:
+                item = output_queue.get(timeout=5)
+                if item is None:
+                    break
+                original_captions, predictions, img_bytes, heights, widths = item
+                predicted_captions = processor.batch_decode(
+                    predictions,
+                    skip_special_tokens=False,
+                )
+                for caption, pred_caption, img_byte, height, width in zip(
+                    original_captions, predicted_captions, img_bytes, heights, widths
+                ):
+                    processed_caption = processor.post_process_generation(
+                        pred_caption, task=prompt, image_size=(width, height)
+                    )[prompt]
+                    original_image = Image.open(io.BytesIO(img_byte)).convert("RGB")
+                    hash_image = insecure_hashlib.sha1(original_image.tobytes()).hexdigest()
+                    img_path = output_dir.joinpath(f"{hash_image}.jpg")
+                    original_image.save(img_path)
+
+                    caption_dict = {"original": caption, "predicted": processed_caption}
+                    with output_dir.joinpath(f"{hash_image}_caption.json").open("w") as f:
+                        json.dump(caption_dict, f, indent=4)
+
+            except queue.Empty:
+                continue
+
+    output_queue = queue.Queue()
+    save_thread = ThreadPoolExecutor(max_workers=num_workers)
+    save_future = save_thread.submit(save_results, output_queue, output_dir, processor)
+
+    try:
+        for _, batch_raw in tqdm(
+            enumerate(dataloader),
+            disable=not distributed_state.is_main_process,
+        ):
+            with distributed_state.split_between_processes(batch_raw) as batch:
+                outputs = model.generate(
+                    input_ids=batch["input_ids"].to(distributed_state.device),
+                    pixel_values=batch["pixel_values"].to(distributed_state.device, model.dtype),
+                    max_new_tokens=max_new_tokens,
+                    num_beams=num_beams,
+                )
+                output_queue.put(
+                    (
+                        batch["original_captions"],
+                        outputs,
+                        batch["img_bytes"],
+                        batch["heights"],
+                        batch["widths"],
+                    )
+                )
+    finally:
+        output_queue.put(None)
+        save_thread.shutdown(wait=True)
+
+    save_future.result()
+
+
+if __name__ == "__main__":
+    fire.Fire(main)
```

Comment:
Would be nice to add some notes above the script about the time it takes to get the artifacts done (maybe on a single shard)?

Reply:
Sure I'll do some benchmarking, any preference on GPU model and count?

Reply:
I think as long as the script runs on 2 GPUs (maybe 4090s), it should be okay. But I will let @muellerzr to comment further.

Reply:
I've tested this on 2x4090. It appears to be CPU bound due to `sha1`, `pil_to_bytes` in preprocess, `Image.open`, `processor` on the batch in collate.

Testing with 1 shard, `https://huggingface.co/datasets/pixparse/cc3m-wds/resolve/main/cc3m-train-0000.tar`, batch size 12 (6 per gpu), and `<CAPTION>` task:
```
420it [10:33,  1.51s/it]
```

Compared to my project [florence-tool](https://github.com/bigdata-pw/florence-tool) using 1 GPU:
```
5046it [06:22, 13.18it/s]
```
In that project I save the results with `__key__` and `__url__`, use `processor` in `preprocess` then concat in `collate` and don't resave the image, although I do need to test whether `processor` is more efficient in `preprocess` or `collate`. Note: I increment the progress bar by batch size rather than batch index.

Testing a version of this script that does the same get roughly the same speed accounting for 1 vs 2 GPUs:
```
420it [03:16,  2.13it/s]
```

WDYT?

Also the last partial batch is missing in this script:
```
ls -f1 outputs/ | grep .json | wc -l
5040
```
Looks like we'd need to use `apply_padding=True` in `split_between_processes`.

Reply:
Thanks!

Not serializing the images is fine. I think it's dependent on how one is ultimately using the datasets in their workflows. But how are we doing the captions? They should be serialized, right? 

What I usually do is first serialize the images and caption, and then package them as `webdataset` shards. I think we could directly do that here instead of serializing the captions. WDYT? 

Additionally, for speeding up the PIL stuff, I find https://github.com/uploadcare/pillow-simd to be quintessential.

Reply:
I've pushed an optimized version that still saves the images and original captions, seems the issue was `pil_to_bytes` and `Image.open`. See the [diff here](https://github.com/huggingface/accelerate/pull/3123/commits/9767e7000594714d2c89a8f6d6d8661df4935565).

Performance:
```
On 2x4090: 420it [03:15,  2.15it/s] (~25.8 images/s)

With --prompt "<DETAILED_CAPTION>": 420it [08:16,  1.18s/it] (~10.17 images/s)
```

Reply:
Nice. IIUC is this same performance you obtained with your tool? 

Additionally, would it be possible to check if pillow-simd leads to any speedup?

Reply:
Yes it's the same performance.

No improvement using pillow-simd.